src/commands/play.rs

```diff
@@ -1,22 +1,156 @@
+    prelude::Mutex,
 };
 
-use songbird::input::Restartable;
-
+use songbird::{input::Restartable, tracks::TrackHandle, Call};
 use youtube_dl::{YoutubeDl, YoutubeDlOutput};
 
-#[command]
-#[aliases("p")]
-async fn play(ctx: &Context, msg: &Message, mut args: Args) -> CommandResult {
+async fn enqueue_song(
+    call: &Arc<Mutex<Call>>,
+    query: String,
+    is_url: bool,
+    flag: &PlayFlag,
+) -> Vec<TrackHandle> {
+    let source = match is_url {
+        true => Restartable::ytdl(query, true).await.unwrap(),
+        false => Restartable::ytdl_search(query, false).await.unwrap(),
+    };
+
+    let mut handler = call.lock().await;
+    handler.enqueue_source(source.into());
+    let mut queue_snapshot = handler.queue().current_queue();
+    drop(handler);
+
+    match flag {
+        PlayFlag::PLAYTOP => {
+            if queue_snapshot.len() > 2 {
+                let handler = call.lock().await;
+
+                handler.queue().modify_queue(|queue| {
+                    let mut not_playing = queue.split_off(1);
+                    not_playing.rotate_right(1);
+                    queue.append(&mut not_playing);
+                });
+
+                queue_snapshot = handler.queue().current_queue();
+            }
+        }
+        _ => (),
+    }
+
+    queue_snapshot
+}
+
+async fn enqueue_playlist(
+    call: &Arc<Mutex<Call>>,
+    uri: &String,
+    flag: &PlayFlag,
+) -> Vec<TrackHandle> {
+    let mut entries = vec![];
+
+    match YoutubeDl::new(uri).flat_playlist(true).run() {
+        Ok(result) => {
+            if let YoutubeDlOutput::Playlist(playlist) = result {
+                entries = playlist.entries.unwrap();
+
+                for entry in &entries {
+                    let uri = format!(
+                        "https://www.youtube.com/watch?v={}",
+                        entry.url.as_ref().unwrap()
+                    );
+                    let source = Restartable::ytdl(uri, true).await.unwrap();
+                    let mut handler = call.lock().await;
+                    handler.enqueue_source(source.into());
+                }
+            }
+        }
+        Err(_) => todo!("Show failed to fetch playlist message!"),
+    }
+
+    let handler = call.lock().await;
+    let mut queue_snapshot = handler.queue().current_queue();
+    drop(handler);
+
+    match flag {
+        PlayFlag::PLAYTOP => {
+            if queue_snapshot.len() > 2 {
+                let handler = call.lock().await;
+
+                handler.queue().modify_queue(|queue| {
+                    let mut not_playing = queue.split_off(1);
+                    let rotations = not_playing.len() - &entries.len();
+                    not_playing.rotate_left(rotations);
+                    queue.append(&mut not_playing);
+                });
+
+                queue_snapshot = handler.queue().current_queue();
+            }
+        }
+        _ => (),
+    }
+
+    queue_snapshot
+}
+
+async fn send_added_to_queue_message(
```

Comment:
Shouldn't this be moved to the same place as `send_simple_message`? Either utils or a new module for message notifications or something.

Reply:
Agreed, moved.